### PR TITLE
NameResolutionPal.Unix enabled cancellation

### PR DESCRIPTION
**Pull request by gfoidl**
#34633 added async support for name resolution on glibc-based Unix-distros, this PR adds cancellation.

### Note

The man-page of [gai_cancel](https://linux.die.net/man/3/gai_cancel) states:

> The request cannot be canceled if it is currently being processed; in that case, it will be handled as if gai_cancel() has never been called.

The [source](https://code.woboq.org/userspace/glibc/resolv/gai_cancel.c.html) also shows that `gai_cancel` just removes the request from the queue, so the behavior might be somewhat unexpected and should be documented appropriate (where?)

### CODE REVIEW

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs [hunk 1]
```diff
@@ -174,19 +174,33 @@ public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
...
-            using var cts = new CancellationTokenSource();
+            const int numberOfRequests = 100;
+
+            using CancellationTokenSource cts = new();
+            List<Task> tasks = new(capacity: numberOfRequests);
 
-            Task task = Dns.GetHostAddressesAsync(TestSettings.UncachedHost, cts.Token);
+            for (int i = 0; i < numberOfRequests; ++i)
+            {
+                Task task = Dns.GetHostAddressesAsync(TestSettings.UncachedHost, cts.Token);
```
Comment (gfoidl) [inline src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:186]: See https://github.com/dotnet/runtime/issues/43816#issuecomment-748088532
If this is consider reliable enough, so that issue is fixed with this change.

File: src/libraries/Native/Unix/System.Native/pal_networking.c [hunk 2]
```diff
@@ -586,7 +590,11 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, int32_t address
...
 
-int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, int32_t addressFamily, HostEntry* entry, GetHostEntryForNameCallback callback)
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address,
+                                              int32_t addressFamily,
+                                              HostEntry* entry,
+                                              GetHostEntryForNameCallback callback,
+                                              void** cancelHandle)
 {
 #if HAVE_GETADDRINFO_A
     if (address == NULL || entry == NULL)
```
Comment (stephentoub) [inline src/libraries/Native/Unix/System.Native/pal_networking.c:600]:
```suggestion
    if (address == NULL || entry == NULL || cancelHandle == NULL)
```
?

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs [hunk 3]
```diff
@@ -279,6 +288,46 @@ public GetHostEntryForNameState(string hostName, bool justAddresses)
...
+
+                lock (this)
+                {
+                    if (_cancellationContext is null)
+                    {
+                        // The operation completed before registration could be done.
+                        return;
+                    }
+
+                    _cancellationTokenRegistration = cancellationToken.UnsafeRegister(o =>
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:303]:
```suggestion
                    _cancellationTokenRegistration = cancellationToken.UnsafeRegister(static o =>
```

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs [hunk 4]
```diff
@@ -119,16 +120,17 @@ private static unsafe void GetHostEntryForNameCallback(Interop.Sys.HostEntry* en
             // Can be casted directly to GetHostEntryForNameContext* because the HostEntry is its first field
             GetHostEntryForNameContext* context = (GetHostEntryForNameContext*)entry;
 
-            ProcessResult(GetSocketErrorForNativeError(error), context);
+            ProcessResult(error, context);
         }
 
-        private static unsafe void ProcessResult(SocketError errorCode, GetHostEntryForNameContext* context)
+        private static unsafe void ProcessResult(int error, GetHostEntryForNameContext* context)
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:126]: Why this change?  We can't continue to use SocketError in the presence of EAI_CANCELED?  Can't it map to one of the existing SocketError values?

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs [hunk 5]
```diff
@@ -174,19 +174,33 @@ public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
         [OuterLoop]
         [Fact]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/43816")] // Race condition outlined below.
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/33378", TestPlatforms.AnyUnix)] // Cancellation of an outstanding getaddrinfo is not supported on *nix.
         public async Task DnsGetHostAddresses_PostCancelledToken_Throws()
         {
-            using var cts = new CancellationTokenSource();
+            const int numberOfRequests = 100;
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:179]: Nit: PascalCasing

Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:179]: Within a method too? Just want to confirm (this is new to me).

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs [hunk 6]
```diff
@@ -174,19 +174,33 @@ public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
...
 
-            OperationCanceledException oce = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => task);
-            Assert.Equal(cts.Token, oce.CancellationToken);
+            await Assert.ThrowsAnyAsync<OperationCanceledException>(async () =>
+            {
+                for (int i = 0; i < numberOfRequests; ++i)
+                {
+                    try
+                    {
+                        await tasks[i].ConfigureAwait(false);
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:198]: ConfigureAwait(false) isn't needed in tests like this.

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs [hunk 7]
```diff
@@ -174,19 +174,33 @@ public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
...
+                    }
+                    catch (Exception ex) when (ex is SocketException)
+                    { }
+                }
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:202]: If the first task is canceled, such that the await throws an OperationCanceledException, this construction won't end up waiting on the rest of the tasks, which is something we strive to avoid in both src and tests (leaving incomplete tasks in flight after the associated operation thinks it's done).  It'd be better to do a WhenAll here, and then process the resulting exceptions, e.g.
```C#
try
{
    await Task.WhenAll(tasks);
}
catch
{
    Assert.Contains(tasks, t => t.Exception?.InnerException is OperationCanceledException); 
}
```
or something like that.

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:186]: I'll let @wfurt comment, but this seems like at some point it's going to fail spuriously in CI.

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:179]: Yup.  `const` == PascalCasing, regardless of where it's used. :)

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs (hunk 4 above)
Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:126]: I didn't find a matching value in https://github.com/dotnet/runtime/blob/f6e73ad8c8a3252a5528724fa241fca410295008/src/libraries/System.Net.Primitives/src/System/Net/Sockets/SocketError.cs#L9

Comment (stephentoub) [reply src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:126]: SocketError.OperationAborted

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs (hunk 7 above)
Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:202]: > leaving incomplete tasks in flight after the associated operation thinks it's done

That's a good point. Thanks.

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs (hunk 4 above)
Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:126]: Ey, so easy...thanks.

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:186]: It seems like the TestSettings.UncachedHost is already used twice and the second use _could_ be cached - certainly in case of systemd resolver. If we want to improve chance of failure, I think each round should use unique name.  And the test was already problematic in CI. 

We could also accept the reality that cancellation is best-effort and timing sensitive and verify that _if_ task failed, OperationCanceledException is only one reason. 

We could also use SkipTestException. That would not change the test run much but we could at least see reports over time.

Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:186]: So we could do this strategy:

1. update the tests here to use SkipTestException, so it won't break CI but we still get some data for the cancellation-tests
1. provide a proper / more robust solution when adressing #43816

Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:186]: 62f05fe25098f96ccba82dc238174ed2a94f4be6 adds `SkipTestException`. OK as proposed in the comment before?

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by karelz** [comment]
@stephentoub any additional concerns, or can we merge?

**Comment by gfoidl** [comment]
Besides the code we should have a look on how to document the somewhat unintuitive behavior of cancellation on unix -- see top post.

**Comment by stephentoub** [comment]
> we should have a look on how to document the somewhat unintuitive behavior of cancellation on unix

All attempts at cancellation are best effort, and there are many places where certain sections of execution aren't interruptible by cancellation.  It's unfortunate that the support provided by glibc isn't better here, but I don't think we need to or even can get into the nitty gritty of exactly what's cancelable when.  Hopefully in the future this can be improved.

**Comment by stephentoub** [comment]
I do have a more general question, though.  Given the limitation, is it worth it?

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs [hunk 8]
```diff
@@ -279,6 +290,46 @@ public GetHostEntryForNameState(string hostName, bool justAddresses)
 
             public Task Task => JustAddresses ? _ipAddressArrayBuilder.Task : _ipHostEntryBuilder.Task;
 
+            public void RegisterForCancellation(CancellationToken cancellationToken)
+            {
+                if (!cancellationToken.CanBeCanceled) return;
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:295]: Nit:
```suggestion
                if (!cancellationToken.CanBeCanceled)
                {
                    return;
                }
```

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs [hunk 9]
```diff
@@ -279,6 +290,46 @@ public GetHostEntryForNameState(string hostName, bool justAddresses)
...
+                    }
+
+                    _cancellationTokenRegistration = cancellationToken.UnsafeRegister(static o =>
+                    {
+                        var self = (GetHostEntryForNameState)o!;
+
+                        lock (self)
+                        {
+                            GetHostEntryForNameContext* context = self._cancellationContext;
+
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:312]: Nit:
```suggestion
```

File: src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs [hunk 10]
```diff
@@ -279,6 +290,46 @@ public GetHostEntryForNameState(string hostName, bool justAddresses)
...
+                    if (_cancellationContext is null)
+                    {
+                        // The operation completed before registration could be done.
+                        return;
+                    }
+
+                    _cancellationTokenRegistration = cancellationToken.UnsafeRegister(static o =>
+                    {
+                        var self = (GetHostEntryForNameState)o!;
+
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs:308]: Nit:
```suggestion
```

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs [hunk 11]
```diff
@@ -172,21 +172,31 @@ public async Task DnsGetHostAddresses_PreCancelledToken_Throws()
...
-            // This test might flake if the cancellation token takes too long to trigger:
-            // It's a race between the DNS server getting back to us and the cancellation processing.
             cts.Cancel();
 
-            OperationCanceledException oce = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => task);
-            Assert.Equal(cts.Token, oce.CancellationToken);
+            try
+            {
+                await Task.WhenAll(tasks);
+                throw new SkipTestException("GetHostAddressesAsync should fail but it did not.");
```
Comment (stephentoub) [inline src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs:195]: No one's going to notice this, especially in CI.  Could we just keep retrying until it does fail?  Folks will notice hangs ;-)

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs (hunk 11 above)
Comment (gfoidl) [reply src/libraries/System.Net.NameResolution/tests/FunctionalTests/GetHostAddressesTest.cs]: Like a148365d8ad1dff4a06e1a28bf29b8a426ad003c?

**Comment by gfoidl** [comment]
> will we often be in the non-cancelable case?

I don't have numbers for this, but my gut feeling tells that with current implemenation in _glibc_, where a request is just removed from the queue, a lot of operations aren't cancelable. Especially https://github.com/dotnet/runtime/issues/43816 wouldn't work ("these tests might be moved into a Docker container with a non-working DNS server that will never respond") as this will just hang.

In e.g. server environment where the queue will be flooded with requests, it's more likeley to actually cancel (= remove from the queue) a request.

TBH I don't know if it is worth it, hope that you have an answer for this 😉

**Comment by wfurt** [comment]
> > will we often be in the non-cancelable case?
> 
> I don't have numbers for this, but my gut feeling tells that with current implemenation in _glibc_, where a request is just removed from the queue, a lot of operations aren't cancelable. Especially #43816 wouldn't work ("these tests might be moved into a Docker container with a non-working DNS server that will never respond") as this will just hang.
> 
> In e.g. server environment where the queue will be flooded with requests, it's more likeley to actually cancel (= remove from the queue) a request.
> 
> TBH I don't know if it is worth it, hope that you have an answer for this 😉

To me, the value of cancelation is ability to fail lookup that takes too long - for whatever reason. That gives caller option to handle that any way they want.  Whether we can actually cancel operation inside glibc is secondary IMHO. Since we register delegate, could we do that somehow?

**Comment by gfoidl** [comment]
> Since we register delegate, could we do that somehow?

[gai_cancel](https://linux.die.net/man/3/gai_cancel) returns either `EAI_CANCELLED`, `EAI_NOTCANCELLED`, or `EAI_ALLDONE`. 
We could return this value to https://github.com/dotnet/runtime/blob/a148365d8ad1dff4a06e1a28bf29b8a426ad003c/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs#L316 and then decide how to continue.

`EAI_CANCELLED` or `EAI_ALLDONE` are already handled by the current state of the PR.

`EAI_NOTCANCELLED` is the problematic case. We could return a canceled task the caller, but would need to keep the context alive in order to get no AV (I assume) if the native lookup completes anytime and invokes the callback https://github.com/dotnet/runtime/blob/a148365d8ad1dff4a06e1a28bf29b8a426ad003c/src/libraries/Native/Unix/System.Native/pal_networking.c#L542 which is https://github.com/dotnet/runtime/blob/a148365d8ad1dff4a06e1a28bf29b8a426ad003c/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs#L384-L387.

Right now I have no real idea on how to do this reliable, withou leaking memory or too much complication of the code.

**Comment by gfoidl** [comment]
Closing as per discussion in https://github.com/dotnet/runtime/issues/48566

In short: `getaddrinfo_a` is a sub-optimal api, that doesn't perform real async name resolution. Other ways will be explored.
